### PR TITLE
Avoid unnecessary copies during protobuf serialization and deserialization

### PR DESCRIPTION
**Pull request by Chilledheart**
- avoid string copy due to google::protobuf::message::SerializeToString
- avoid string copy due to google::protobuf::message::ParseFromString

> I thought I can reopen #967 which was closed by mistake. But once I changed the commit, there is no way to reopen it again. Sorry for the inconvenience.

### CODE REVIEW

File: src/cpp/proto/proto_utils.cc [hunk 1]
```diff
@@ -36,37 +36,79 @@
...
+  }
+  ~GrpcBufferReader() GRPC_OVERRIDE {
+    grpc_byte_buffer_reader_destroy(reader_);
+  }
+  bool Next(const void **data, int *size) GRPC_OVERRIDE {
+    if (backup_count_ > 0) {
+      *data = GPR_SLICE_START_PTR(slice_) + GPR_SLICE_LENGTH(slice_) -
+              backup_count_;
+      *size = backup_count_;
+      backup_count_ = 0;
```
Comment (yang-g) [inline src/cpp/proto/proto_utils.cc]: Update byte_count_ here? Or maybe return byte_count_-backup_count_ in ByteCount.

BTW. Please format code according to @ctiller's comments in the previous PR.

Comment (Chilledheart) [inline src/cpp/proto/proto_utils.cc]: Updated. Sorry, I used LLVM Style clang-format everywhere and forgot changing for Google-Style option. It should be formatted correctly now.

File: src/cpp/proto/proto_utils.cc [hunk 2]
```diff
@@ -35,38 +35,136 @@
...
+    *data = GPR_SLICE_START_PTR(slice_);
+    byte_count_ += *size = GPR_SLICE_LENGTH(slice_);
+    return true;
   }
-  grpc_byte_buffer_reader_destroy(reader);
-  return msg->ParseFromString(msg_string);
+
+  void BackUp(int count) GRPC_OVERRIDE {
+    backup_count_ = count;
+    byte_count_ -= count;
```
Comment (yang-g) [inline src/cpp/proto/proto_utils.cc]: This should be removed when you have line 148.

Comment (Chilledheart) [inline src/cpp/proto/proto_utils.cc]: Removed in the latest patch.

**Pull request merged**